### PR TITLE
Add helpful methods for Square and Direction

### PR DESCRIPTION
**Pull request by ViliamVadocz**
`checked_step` for moving along a direction while checking whether we have left the board.

`rotate` and `mirror` are useful for generating symmetries.

I also included some tests.

### CODE REVIEW

File: src/ptn.rs [hunk 1]
```diff
@@ -105,6 +105,48 @@ impl FromStr for Square {
     }
 }
 
+impl Square {
```
Comment (alion02) [inline src/ptn.rs]: Move the implementation to the previous impl block (both instances)

File: src/ptn.rs [hunk 2]
```diff
@@ -105,6 +105,48 @@ impl FromStr for Square {
...
+            Down => (column, row.checked_add(1).filter(|&y| y < board_size)?),
+            Left => (column.checked_sub(1)?, row),
+            Right => (column.checked_add(1).filter(|&x| x < board_size)?, row),
+        };
+        Some(Self { column, row })
+    }
+
+    /// Rotate 1 quarter turn counterclockwise.
+    #[must_use]
+    pub const fn rotate(&self, board_size: u8) -> Self {
```
Comment (alion02) [inline src/ptn.rs]: Use pass by move for consistency (all instances)

File: src/ptn.rs [hunk 3]
```diff
@@ -105,6 +105,48 @@ impl FromStr for Square {
...
+
+    /// Generate a Square that is one step in the direction from this Square.
+    /// If the generated Square is outside of the board, then this returns None instead.
+    pub fn checked_step(self, direction: Direction, board_size: u8) -> Option<Self> {
+        self.assert_on_board(board_size);
+        use Direction::*;
+        let (column, row) = (self.column, self.row);
+        let (column, row) = match direction {
+            Up => (column, row.checked_sub(1)?),
+            Down => (column, row.checked_add(1).filter(|&y| y < board_size)?),
```
Comment (alion02) [inline src/ptn.rs]: Up is +y, down is -y; also check rotations

**Review by alion02 — CHANGES_REQUESTED** [review]

File: src/ptn.rs (hunk 3 above)
Comment (ViliamVadocz) [reply src/ptn.rs]: Good catch. This was actually a mistake.
Rotations are correct.

File: src/ptn.rs [hunk 4]
```diff
@@ -40,6 +40,46 @@ impl Square {
...
+        let (column, row) = match direction {
+            Up => (column, row.checked_add(1).filter(|&y| y < board_size)?),
+            Down => (column, row.checked_sub(1)?),
+            Left => (column.checked_sub(1)?, row),
+            Right => (column.checked_add(1).filter(|&x| x < board_size)?, row),
+        };
+        Some(Self { column, row })
+    }
+
+    /// Rotate 1 quarter turn counterclockwise.
```
Comment (alion02) [inline src/ptn.rs]: Docs disagree with tests. Also, it would be inconsistent to rotate Squares clockwise and Directions counterclockwise.

**Review by alion02 — CHANGES_REQUESTED** [review]

File: src/ptn.rs (hunk 4 above)
Comment (ViliamVadocz) [reply src/ptn.rs]: You're right. This is related to the y up/down thing. In my original code I had 0,0 in the top left with increasing y meaning doing down. Switching that makes these rotations clockwise. I fixed the Direction rotation to reflect this as well.

**Review by alion02 — APPROVED** [review]

**Pull request merged**